File: 백준/Silver/15657. N과 M （8）/N과 M （8）.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [N, M] = input.shift().split(" ").map(Number);

input = input[0]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);

const sequence = [];

function dfs(depth) {
  if (depth === M) {
    console.log(sequence.join(" "));
    return;
  }

  for (let i = 0; i < N; i++) {
    if (sequence[sequence.length - 1] > input[i]) continue;
    sequence.push(input[i]);
    dfs(depth + 1);
    sequence.pop();
  }
}

dfs(0);

File: 백준/Silver/15657. N과 M （8）/N과 M （8）.py

N, M = map(int, input().split())
nums = list(map(int, input().split()))
stack = []
# visited = [0] * (N+1)

nums.sort()


def dfs(depth):
    if depth == M:
        print(*stack)
        return

    for idx, num in enumerate(nums):
        if stack and stack[-1] > num:
            continue
        stack.append(num)
        dfs(depth+1)
        stack.pop()


dfs(0)
